### PR TITLE
ci: fix flaky test_pyobject_drop_without_gil_doesnt_decrease_refcnt

### LINKED ISSUE #2723: Flaky test `test_pyobject_drop_without_gil_doesnt_decrease_refcnt`
```
---- gil::tests::test_pyobject_drop_without_gil_doesnt_decrease_refcnt stdout ----
thread 'gil::tests::test_pyobject_drop_without_gil_doesnt_decrease_refcnt' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `2`', src/gil.rs:619:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:483
   1: std::panicking::begin_panic_fmt
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/std/src/panicking.rs:437
   2: pyo3::gil::tests::test_pyobject_drop_without_gil_doesnt_decrease_refcnt
             at ./src/gil.rs:619
   3: pyo3::gil::tests::test_pyobject_drop_without_gil_doesnt_decrease_refcnt::{{closure}}
             at ./src/gil.rs:601
   4: core::ops::function::FnOnce::call_once
             at /home/runner/.rustup/toolchains/1.48.0-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:227
   5: core::ops::function::FnOnce::call_once
             at /rustc/7eac88abb2e57e752f3302f02be5f3ce3d7adfb4/library/core/src/ops/function.rs:227
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

https://github.com/PyO3/pyo3/actions/runs/3402526740/jobs/5658406562

### PR DESCRIPTION
**Pull request by davidhewitt**
Closes #2723 

I think @adamreichold is correct in what causes the flakiness, and the solution in this case seems straightforward enough - just pass `reference` off to another thread to drop it (as that thread won't have the GIL), no events needed.

### CODE REVIEW

**Review by adamreichold — APPROVED** [review]
The solution of just never releasing the GIL is much tidier than what I anticipated in #2723. Very nice!

bors r+

**Comment by adamreichold** [comment]
Yet Clippy can never be appeased...

**Comment by davidhewitt** [comment]
bors r=adamreichold

File: src/gil.rs [hunk 1]
```diff
@@ -577,54 +588,51 @@ mod tests {
...
+        let obj = Python::with_gil(|py| {
+            let obj = get_object(py);
+            // Create a reference to drop without the GIL.
+            let reference = obj.clone_ref(py);
+
+            assert_eq!(obj.get_refcnt(py), 2);
+            assert!(pool_not_dirty());
+
+            // Drop reference in a separate thread which doesn't have the GIL.
+            std::thread::spawn(move || drop(reference)).join().unwrap();
```
Comment (adamreichold) [inline src/gil.rs:619]: I suspect this test function needs a

```rust
#[cfg(not(target_arch = "wasm32"))] // We are building wasm Python with pthreads disabled
```

attribute like the other tests which spawn threads.

**Comment by adamreichold** [comment]
(I suspect bors does not listen to me since I could not push to the protected branch any way, but let's try nevertheless.)

bors r+

**Comment by davidhewitt** [comment]
> I suspect bors does not listen to me

Hmm looks like just worked here, can check its config if you think there's an issue?

Thanks for fixing up this one 😁

**Pull request merged**